simple_parsing/decorators.py

```diff
@@ -0,0 +1,122 @@
+import collections
+import dataclasses
+import functools
+import inspect
+from typing import Any, Callable, NamedTuple, Optional, Type
+
+import docstring_parser as dp
+
+from . import helpers, parsing
+
+
+class _Field(NamedTuple):
+    name: str
+    annotation: Type
+    field: dataclasses.Field
+
+
+def main(original_function: Optional[Callable] = None, **sp_kwargs) -> Callable:
+    """Parse a function's arguments using simple-parsing from type annotations."""
+
+    def _description_from_docstring(docstring: dp.Docstring) -> str:
+        """Construct a description from the short and long description of a docstring."""
+        description = ""
+        if docstring.short_description:
+            description += f"{docstring.short_description}\n"
+            if docstring.blank_after_short_description:
+                description += "\n"
+        if docstring.long_description:
+            description += f"{docstring.long_description}\n"
+            if docstring.blank_after_long_description:
+                description += "\n"
+        return description
+
+    def _decorate_with_cli_args(function: Callable) -> Callable:
+        """Decorate `function` by binding its arguments obtained from simple-parsing."""
+
+        @functools.wraps(function)
+        def _wrapper(*other_args, **other_kwargs) -> Any:
+            # Parse signature and parameters
+            signature = inspect.signature(function, follow_wrapped=True)
+            parameters = signature.parameters
+
+            # Parse docstring to use as help strings
+            docstring = dp.parse(inspect.getdoc(function) or "")
+            docstring_param_description = {
+                param.arg_name: param.description for param in docstring.params
+            }
+
+            # Parse all arguments from the function
+            fields = []
+            for name, parameter in parameters.items():
+                # Replace empty annotation with Any
+                if parameter.annotation == inspect.Parameter.empty:
+                    parameter = parameter.replace(annotation=Any)
+
+                # Parse default or default_factory if the default is callable.
+                default, default_factory = dataclasses.MISSING, dataclasses.MISSING
+                if parameter.default != inspect.Parameter.empty:
+                    if inspect.isfunction(parameter.default):
+                        default_factory = parameter.default
+                    else:
+                        default = parameter.default
+
+                field = _Field(
+                    name,
+                    parameter.annotation,
```

Comment:
What happens if the user script is using postponed evaluation of type annotations (`from __future__ import annotations`)? I assume that `parameter.annotation` would then be a string. Would this code be affected by that?

The rest of simple-parsing can deal fairly well with postponed evaluations, by evaluating them back into a runtime type annotations (e.g. `"int | bool"` gets brought back down to a `typing.Union[int, bool]` later down the line, you don't have to worry about doing that here)

Reply:
Re: We can take a look at postponed evaluations later. This should be fine for now.